File: service/web_fetch/contracts/src/envelope.rs
//! The reusable tool-result envelope, version 1.
//!
//! Webfetch is its first user, not its owner: a later consumer (research is the
//! one the plan names) declares its own [`Payload`] and reuses every control
//! field here unchanged. That is what keeps a second tool from inventing a
//! second way to say "this came from outside and was checked".

use serde::{Deserialize, Serialize, de::DeserializeOwned};

use crate::{ContractError, Result};

/// The only envelope version this build understands.
pub const ENVELOPE_VERSION: u32 = 1;

/// A typed tool payload carried in [`Envelope::data`].
///
/// Implementors own their own field validation; the envelope owns the control
/// fields and refuses to carry a payload under the wrong tool identity.
pub trait Payload: Serialize + DeserializeOwned {
    /// Registered tool key, e.g. `web_fetch`.
    const TOOL: &'static str;
    /// Registered tool version, e.g. `1`.
    const TOOL_VERSION: u32;

    /// Validate payload-internal invariants.
    fn validate(&self) -> Result<()>;
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Status {
    Ok,
    Rejected,
    Error,
}

/// What the result says about the data's provenance. It describes bytes; it
/// grants nothing.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Trust {
    ExternalUntrusted,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum InspectionStatus {
    /// Every required check completed and none matched.
    NoMatch,
    /// At least one enabled rule matched.
    Match,
    /// A check could not be completed: crash, timeout, partial scan, no bundle.
    Failed,
    /// The required checks never started, because the response never got that far.
    NotRun,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct Inspection {
    pub status: InspectionStatus,
    /// Reference to the immutable policy that produced `status`. Null only when
    /// no policy was in force, which can never accompany a success.
    pub policy_id: Option<String>,
}

impl Inspection {
    pub fn new(status: InspectionStatus, policy_id: Option<String>) -> Self {
        Self { status, policy_id }
    }

    pub fn not_run() -> Self {
        Self::new(InspectionStatus::NotRun, None)
    }

    /// The inspection started and could not be completed. Pair this with
    /// [`ErrorCode::InspectionFailed`]; `not_run` is not a valid companion for
    /// that code, because it would claim the checks never began.
    pub fn failed(policy_id: Option<String>) -> Self {
        Self::new(InspectionStatus::Failed, policy_id)
    }
}

/// Fixed failure codes. Every one maps to a constant message, so no code path
/// can smuggle remote text, a URL, a matched string, or a worker backtrace into
/// an error the model will read.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ErrorCode {
    /// An enabled rule matched anywhere in the response.
    ContentRejected,
    /// The inspection could not be completed.
    InspectionFailed,
    /// A destination or redirect target was not allowed.
    UnsafeSource,
    /// Unsupported media type, charset, or malformed encoding.
    UnsupportedContent,
    /// A header, wire-byte, decoded-byte, or output limit was exceeded.
    ResponseLimitExceeded,
    /// The request could not be completed at the network level.
    FetchFailed,
    /// The whole-run deadline elapsed.
    DeadlineExceeded,
    /// The stage or result gate failed in a way that is not the response's fault.
    InternalError,
}

impl ErrorCode {
    /// The one message this code is ever allowed to carry.
    pub const fn message(self) -> &'static str {
        match self {
            Self::ContentRejected => "Response content was withheld from model context.",
            Self::InspectionFailed => {
                "Response inspection did not complete; no content was returned."
            }
            Self::UnsafeSource => "The requested destination is not allowed.",
            Self::UnsupportedContent => "The response media type or encoding is not supported.",
            Self::ResponseLimitExceeded => "The response exceeded a configured limit.",
            Self::FetchFailed => "The request to the source could not be completed.",
            Self::DeadlineExceeded => "The request did not complete within its deadline.",
            Self::InternalError => "The tool could not complete this request.",
        }
    }

    /// Inspection statuses this code may be reported with.
    ///
    /// `no_match` is absent everywhere: a completed clean inspection is only
    /// ever reported alongside returned content, so a failure can never claim
    /// the response was checked and found clean.
    fn permits(self, status: InspectionStatus) -> bool {
        match self {
            Self::ContentRejected => status == InspectionStatus::Match,
            Self::InspectionFailed => status == InspectionStatus::Failed,
            _ => matches!(status, InspectionStatus::NotRun | InspectionStatus::Failed),
        }
    }

    const fn status(self) -> Status {
        match self {
            Self::ContentRejected => Status::Rejected,
            _ => Status::Error,
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct ToolError {
    pub code: ErrorCode,
    pub message: String,
}

impl ToolError {
    /// The only constructor: the message comes from the code, never from a caller.
    pub fn new(code: ErrorCode) -> Self {
        Self {
            code,
            message: code.message().to_string(),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct Envelope<P> {
    pub envelope_version: u32,
    pub tool: String,
    pub tool_version: u32,
    pub status: Status,
    pub trust: Trust,
    pub inspection: Inspection,
    pub data: Option<P>,
    pub error: Option<ToolError>,
}

impl<P: Payload> Envelope<P> {
    /// Build a success. Callers cannot assemble one without a completed clean
    /// inspection and the policy that produced it.
    pub fn ok(data: P, policy_id: impl Into<String>) -> Result<Self> {
        let envelope = Self {
            envelope_version: ENVELOPE_VERSION,
            tool: P::TOOL.to_string(),
            tool_version: P::TOOL_VERSION,
            status: Status::Ok,
            trust: Trust::ExternalUntrusted,
            inspection: Inspection::new(InspectionStatus::NoMatch, Some(policy_id.into())),
            data: Some(data),
            error: None,
        };
        envelope.validate()?;
        Ok(envelope)
    }

    /// Build a rejection or error. The status follows from the code.
    pub fn failure(code: ErrorCode, inspection: Inspection) -> Result<Self> {
        let envelope = Self {
            envelope_version: ENVELOPE_VERSION,
            tool: P::TOOL.to_string(),
            tool_version: P::TOOL_VERSION,
            status: code.status(),
            trust: Trust::ExternalUntrusted,
            inspection,
            data: None,
            error: Some(ToolError::new(code)),
        };
        envelope.validate()?;
        Ok(envelope)
    }

    /// Parse and validate an envelope of this payload type.
    ///
    /// Consumers validate before admitting anything to model context, so this is
    /// the same gate on both sides of the wire.
    pub fn from_json(value: &str) -> Result<Self> {
        let envelope: Self = serde_json::from_str(value)
            .map_err(|error| ContractError::new(format!("envelope is not valid: {error}")))?;
        envelope.validate()?;
        Ok(envelope)
    }

    pub fn to_json(&self) -> Result<String> {
        self.validate()?;
        serde_json::to_string(self).map_err(|error| {
            ContractError::new(format!("envelope could not be serialized: {error}"))
        })
    }

    pub fn validate(&self) -> Result<()> {
        if self.envelope_version != ENVELOPE_VERSION {
            return Err(ContractError::new(format!(
                "unsupported envelope_version: {}",
                self.envelope_version
            )));
        }
        if self.tool != P::TOOL || self.tool_version != P::TOOL_VERSION {
            return Err(ContractError::new(format!(
                "envelope carries {}:v{}, expected {}:v{}",
                self.tool,
                self.tool_version,
                P::TOOL,
                P::TOOL_VERSION
            )));
        }
        match (self.status, &self.data, &self.error) {
            (Status::Ok, Some(data), None) => {
                if self.inspection.status != InspectionStatus::NoMatch {
                    return Err(ContractError::new(
                        "ok requires a completed no_match inspection",
                    ));
                }
                if self.inspection.policy_id.is_none() {
                    return Err(ContractError::new(
                        "ok requires the policy that cleared the response",
                    ));
                }
                data.validate()?;
            }
            (Status::Ok, _, _) => {
                return Err(ContractError::new(
                    "ok requires non-null data and null error",
                ));
            }
            (status, None, Some(error)) => {
                if error.code.status() != status {
                    return Err(ContractError::new(format!(
                        "{:?} is not the status for error code {:?}",
                        status, error.code
                    )));
                }
                if error.message != error.code.message() {
                    return Err(ContractError::new(
                        "error.message is not this code's fixed message",
                    ));
                }
                if !error.code.permits(self.inspection.status) {
                    return Err(ContractError::new(format!(
                        "inspection status {:?} cannot accompany error code {:?}",
                        self.inspection.status, error.code
                    )));
                }
            }
            _ => {
                return Err(ContractError::new(
                    "a failure requires null data and a fixed error",
                ));
            }
        }
        if let Some(policy_id) = &self.inspection.policy_id {
            validate_policy_id(policy_id)?;
        } else if self.inspection.status == InspectionStatus::Match {
            return Err(ContractError::new(
                "a match must name the policy that matched",
            ));
        }
        Ok(())
    }
}

/// Policy IDs are server-owned identifiers, not free text copied from anywhere.
fn validate_policy_id(value: &str) -> Result<()> {
    let shaped = !value.is_empty()
        && value.len() <= 64
        && value.chars().all(|character| {
            character.is_ascii_lowercase()
                || character.is_ascii_digit()
                || matches!(character, '-' | '_' | '.')
        });
    if shaped {
        Ok(())
    } else {
        Err(ContractError::new(
            "inspection.policy_id is not a valid policy identifier",
        ))
    }
}

File: service/ai_jobs/src/ai_jobs/envelope.py
"""The reusable tool-result envelope, version 1.

The Rust stages build these; the control plane validates them before anything
reaches a caller, and a future consumer validates them before admitting content
to model context. Both sides enforce the same rules, and the shared fixtures in
``service/web_fetch/fixtures`` are what keeps them agreeing.

Two invariants are worth stating plainly, because they are the reason the
envelope exists rather than a bare payload:

* Control fields are server-owned. Source text that looks like an envelope is
  still source text; it arrives inside ``data`` and cannot set ``status``,
  ``trust``, or the tool identity of the envelope carrying it.
* A failure carries no remote text. Each code has exactly one message, and a
  value whose message is anything else is not a valid envelope.
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import StrEnum
from typing import Any, Callable, Mapping

from ai_jobs.contracts import (
    ContractError,
    require_exact_fields,
    require_object,
    require_string,
)

ENVELOPE_VERSION = 1


class Status(StrEnum):
    OK = "ok"
    REJECTED = "rejected"
    ERROR = "error"


class Trust(StrEnum):
    """Describes fetched data. It grants nothing."""

    EXTERNAL_UNTRUSTED = "external_untrusted"


class InspectionStatus(StrEnum):
    NO_MATCH = "no_match"
    MATCH = "match"
    FAILED = "failed"
    NOT_RUN = "not_run"


class ErrorCode(StrEnum):
    CONTENT_REJECTED = "content_rejected"
    INSPECTION_FAILED = "inspection_failed"
    UNSAFE_SOURCE = "unsafe_source"
    UNSUPPORTED_CONTENT = "unsupported_content"
    RESPONSE_LIMIT_EXCEEDED = "response_limit_exceeded"
    FETCH_FAILED = "fetch_failed"
    DEADLINE_EXCEEDED = "deadline_exceeded"
    INTERNAL_ERROR = "internal_error"


#: The one message each code may carry. Mirrors ``ErrorCode::message`` in the
#: Rust contracts crate.
ERROR_MESSAGES: Mapping[ErrorCode, str] = {
    ErrorCode.CONTENT_REJECTED: "Response content was withheld from model context.",
    ErrorCode.INSPECTION_FAILED: "Response inspection did not complete; no content was returned.",
    ErrorCode.UNSAFE_SOURCE: "The requested destination is not allowed.",
    ErrorCode.UNSUPPORTED_CONTENT: "The response media type or encoding is not supported.",
    ErrorCode.RESPONSE_LIMIT_EXCEEDED: "The response exceeded a configured limit.",
    ErrorCode.FETCH_FAILED: "The request to the source could not be completed.",
    ErrorCode.DEADLINE_EXCEEDED: "The request did not complete within its deadline.",
    ErrorCode.INTERNAL_ERROR: "The tool could not complete this request.",
}

#: Only ``content_rejected`` is a rejection; everything else is a tool error.
_REJECTION_CODES = frozenset({ErrorCode.CONTENT_REJECTED})

#: Inspection statuses each code may be reported with. ``no_match`` appears
#: nowhere: a completed clean inspection is only ever reported with content, so
#: no failure can claim the response was checked and found clean.
_PERMITTED_INSPECTION: Mapping[ErrorCode, frozenset[InspectionStatus]] = {
    ErrorCode.CONTENT_REJECTED: frozenset({InspectionStatus.MATCH}),
    ErrorCode.INSPECTION_FAILED: frozenset({InspectionStatus.FAILED}),
}
_DEFAULT_PERMITTED_INSPECTION = frozenset({InspectionStatus.NOT_RUN, InspectionStatus.FAILED})


def status_for(code: ErrorCode) -> Status:
    return Status.REJECTED if code in _REJECTION_CODES else Status.ERROR


@dataclass(frozen=True, slots=True)
class Inspection:
    status: InspectionStatus
    policy_id: str | None

    def to_dict(self) -> dict[str, object]:
        return {"status": self.status.value, "policy_id": self.policy_id}

    @classmethod
    def from_dict(cls, value: object) -> "Inspection":
        data = require_object(value, "inspection")
        require_exact_fields(data, {"status", "policy_id"}, "inspection")
        try:
            status = InspectionStatus(require_string(data["status"], "inspection.status", 32))
        except ValueError as exc:
            raise ContractError("inspection.status is not a known inspection status") from exc
        policy_id = data["policy_id"]
        if policy_id is not None:
            policy_id = require_string(policy_id, "inspection.policy_id", 64)
            if not _is_policy_id(policy_id):
                raise ContractError("inspection.policy_id is not a valid policy identifier")
        return cls(status=status, policy_id=policy_id)


@dataclass(frozen=True, slots=True)
class EnvelopeError:
    code: ErrorCode
    message: str

    @classmethod
    def for_code(cls, code: ErrorCode) -> "EnvelopeError":
        """The only way to build one: the message comes from the code."""
        return cls(code=code, message=ERROR_MESSAGES[code])

    def to_dict(self) -> dict[str, str]:
        return {"code": self.code.value, "message": self.message}

    @classmethod
    def from_dict(cls, value: object) -> "EnvelopeError":
        data = require_object(value, "error")
        require_exact_fields(data, {"code", "message"}, "error")
        try:
            code = ErrorCode(require_string(data["code"], "error.code", 64))
        except ValueError as exc:
            raise ContractError("error.code is not a public error code") from exc
        message = require_string(data["message"], "error.message", 500)
        if message != ERROR_MESSAGES[code]:
            raise ContractError("error.message is not this code's fixed message")
        return cls(code=code, message=message)


#: Validates a tool's payload. Raises ContractError if the payload is not valid
#: for that tool and version.
PayloadValidator = Callable[[object], Any]


@dataclass(frozen=True, slots=True)
class Envelope:
    envelope_version: int
    tool: str
    tool_version: int
    status: Status
    trust: Trust
    inspection: Inspection
    data: Any | None
    error: EnvelopeError | None

    @classmethod
    def ok(cls, tool: str, tool_version: int, data: Any, policy_id: str) -> "Envelope":
        return cls(
            envelope_version=ENVELOPE_VERSION,
            tool=tool,
            tool_version=tool_version,
            status=Status.OK,
            trust=Trust.EXTERNAL_UNTRUSTED,
            inspection=Inspection(InspectionStatus.NO_MATCH, policy_id),
            data=data,
            error=None,
        )

    @classmethod
    def failure(cls, tool: str, tool_version: int, code: ErrorCode, inspection: Inspection) -> "Envelope":
        envelope = cls(
            envelope_version=ENVELOPE_VERSION,
            tool=tool,
            tool_version=tool_version,
            status=status_for(code),
            trust=Trust.EXTERNAL_UNTRUSTED,
            inspection=inspection,
            data=None,
            error=EnvelopeError.for_code(code),
        )
        _check(envelope, None)
        return envelope

    @classmethod
    def from_dict(
        cls,
        value: object,
        *,
        tool: str,
        tool_version: int,
        payload: PayloadValidator | None = None,
    ) -> "Envelope":
        """Parse and validate an envelope for one named tool and version."""
        data = require_object(value, "envelope")
        require_exact_fields(
            data,
            {"envelope_version", "tool", "tool_version", "status", "trust", "inspection", "data", "error"},
            "envelope",
        )
        version = data["envelope_version"]
        if isinstance(version, bool) or not isinstance(version, int):
            raise ContractError("envelope.envelope_version must be an integer")
        if version != ENVELOPE_VERSION:
            raise ContractError(f"unsupported envelope_version: {version}")
        tool_key = require_string(data["tool"], "envelope.tool", 64)
        carried_version = data["tool_version"]
        if isinstance(carried_version, bool) or not isinstance(carried_version, int):
            raise ContractError("envelope.tool_version must be an integer")
        if tool_key != tool or carried_version != tool_version:
            raise ContractError(
                f"envelope carries {tool_key}:v{carried_version}, expected {tool}:v{tool_version}"
            )
        try:
            status = Status(require_string(data["status"], "envelope.status", 32))
            trust = Trust(require_string(data["trust"], "envelope.trust", 32))
        except ValueError as exc:
            raise ContractError("envelope status or trust is not a known value") from exc
        inspection = Inspection.from_dict(data["inspection"])
        error = None if data["error"] is None else EnvelopeError.from_dict(data["error"])
        envelope = cls(
            envelope_version=version,
            tool=tool_key,
            tool_version=carried_version,
            status=status,
            trust=trust,
            inspection=inspection,
            data=data["data"],
            error=error,
        )
        _check(envelope, payload)
        return envelope

    def to_dict(self) -> dict[str, object]:
        data = self.data
        if data is not None and hasattr(data, "to_dict"):
            data = data.to_dict()
        return {
            "envelope_version": self.envelope_version,
            "tool": self.tool,
            "tool_version": self.tool_version,
            "status": self.status.value,
            "trust": self.trust.value,
            "inspection": self.inspection.to_dict(),
            "data": data,
            "error": None if self.error is None else self.error.to_dict(),
        }


def _check(envelope: Envelope, payload: PayloadValidator | None) -> None:
    if envelope.status is Status.OK:
        if envelope.data is None or envelope.error is not None:
            raise ContractError("ok requires non-null data and null error")
        if envelope.inspection.status is not InspectionStatus.NO_MATCH:
            raise ContractError("ok requires a completed no_match inspection")
        if envelope.inspection.policy_id is None:
            raise ContractError("ok requires the policy that cleared the response")
        if payload is not None:
            payload(envelope.data)
        return

    if envelope.data is not None or envelope.error is None:
        raise ContractError("a failure requires null data and a fixed error")
    if status_for(envelope.error.code) is not envelope.status:
        raise ContractError(f"{envelope.status.value} is not the status for {envelope.error.code.value}")
    permitted = _PERMITTED_INSPECTION.get(envelope.error.code, _DEFAULT_PERMITTED_INSPECTION)
    if envelope.inspection.status not in permitted:
        raise ContractError(
            f"inspection status {envelope.inspection.status.value} cannot accompany {envelope.error.code.value}"
        )
    if envelope.inspection.status is InspectionStatus.MATCH and envelope.inspection.policy_id is None:
        raise ContractError("a match must name the policy that matched")


def _is_policy_id(value: str) -> bool:
    """Server-owned identifier, ASCII only - the same rule the Rust side applies."""
    return bool(value) and all(
        ("a" <= character <= "z") or ("0" <= character <= "9") or character in "-_." for character in value
    )
